src/mug/tls.gleam

```diff
@@ -0,0 +1,630 @@
+///
+/// The socket is created in passive mode, meaning the the `receive` function is
+/// to be called to receive packets from the client. The
+/// `receive_next_packet_as_message` function can be used to switch the socket
+/// to active mode and receive the next packet as an Erlang message.
+///
+pub fn connect(options: TlsConnectionOptions) -> Result(Socket, Error) {
+  let host = charlist.from_string(options.host)
+  ssl_connect(
+    host,
+    options.port,
+    get_tls_options(options.certificates),
+    options.timeout,
+  )
+}
+
+/// Upgrade a plain TCP connection to TLS.
+///
+/// Accepts a TCP socket and performs the client-side TLS handshake. This
+/// may not work on all TCP servers.
+///
+/// Returns an error if the connection could not be established.
+///
+/// The socket is created in passive mode, meaning the the `receive` function is
+/// to be called to receive packets from the client. The
+/// `receive_next_packet_as_message` function can be used to switch the socket
+/// to active mode and receive the next packet as an Erlang message.
+///
+/// The `certs_keys` option can be 
+///
+pub fn upgrade(
+  socket: TcpSocket,
+  certificates certificates: Certificates,
+  milliseconds timeout: Int,
+) -> Result(Socket, Error) {
+  ssl_upgrade(socket, get_tls_options(certificates), timeout)
+}
+
+/// Whether to use system certificates or not.
+pub fn use_system_cacerts(system: Bool) {
+  Certificates(system, None, [])
+}
+
+fn get_tls_options(
+  certificates: Certificates,
+) -> List(#(TlsOptionName, Dynamic)) {
+  [
+    // When data is received on the socket queue it in the TCP stack rather than
+    // sending it as an Erlang message to the socket owner's inbox.
+    #(Active, dynamic.from(False)),
+    // We want the data from the socket as bit arrays please, not lists.
+    #(Mode, dynamic.from(Binary)),
+    #(
+      Verify,
+      dynamic.from(case certificates {
+        NoVerification -> VerifyNone
+        _ -> VerifyPeer
+      }),
+    ),
+  ]
+  |> fn(opts) {
+    case certificates {
+      NoVerification -> [#(Verify, dynamic.from(VerifyNone)), ..opts]
+      Certificates(system, cacerts: cacerts, certs_keys: certs_keys) -> [
+        #(Verify, dynamic.from(VerifyPeer)),
+        get_cacerts(system, cacerts),
+        #(CertsKeys, dynamic.from(list.map(certs_keys, certs_keys_to_erl))),
+        ..opts
+      ]
+    }
+  }
+}
+
+fn get_cacerts(system: Bool, cacerts: Option(CaCertificates)) -> TlsOption {
+  case system, cacerts {
+    False, Some(DerEncodedCaCertificates(cacerts)) -> #(
+      Cacerts,
+      dynamic.from(cacerts),
+    )
+    True, Some(DerEncodedCaCertificates(cacerts)) -> #(
+      Cacerts,
+      dynamic.from(list.concat([get_system_cacerts().0, cacerts])),
+    )
+    _, Some(PemEncodedCaCertificates(cacerts)) -> #(
+      Cacertfile,
+      dynamic.from(string.to_utf_codepoints(cacerts)),
+    )
+    True, None -> #(Cacerts, dynamic.from(get_system_cacerts()))
```

Comment:
Why do we load the certs rather than have the SSL application do that? It would be able to cache them, for example.

Reply:
I don't quite understand what you mean. Does the SSL application automatically load system cacerts?

Reply:
And caching is done by erlang's ssl module automatically